### PR TITLE
add huggingface classification

### PR DESCRIPTION
**Pull request by Panzy-18**
+ add config and notebook for huggingface classification
+ fix nit in bedrock classification

### CODE REVIEW

File: uniflow/flow/config.py [hunk 1]
```diff
@@ -328,7 +328,6 @@ class RaterForClassificationBedrockClaudeConfig(RaterConfig):
                                                answer, label, and explanation for each case.
     """
 
-    flow_name: str = "RaterFlow"
```
Comment (CambioML) [inline uniflow/flow/config.py:331]: qq: is there a reason to remove the default value for `flow_name` while require user to must pass it in?

File: example/rater/huggingface_classification.ipynb [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+    "%reload_ext autoreload\n",
+    "%autoreload 2\n",
+    "\n",
+    "import sys\n",
+    "\n",
+    "sys.path.append(\".\")\n",
+    "sys.path.append(\"..\")\n",
+    "sys.path.append(\"../..\")\n",
+    "\n",
+    "!{sys.executable} -m pip install transformers accelerate bitsandbytes scipy"
```
Comment (CambioML) [inline example/rater/huggingface_classification.ipynb:86]: let's do `!{sys.executable} -m pip install -q transformers accelerate bitsandbytes scipy` for quite install without verbose logs.

File: example/rater/bedrock_classification.ipynb [hunk 3]
```diff
@@ -376,13 +376,6 @@
     "pprint.pprint(output)"
    ]
   },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "We can see that model response is a json object."
-   ]
-  },
```
[the remark this answers is not in the document]
Comment (Panzy-18) [reply example/rater/bedrock_classification.ipynb:385]: This is the output in this notebook is not JSON format. In fact, I think only GPT3.5/4 support json object mode.

File: uniflow/flow/config.py (hunk 1 above)
Comment (Panzy-18) [reply uniflow/flow/config.py:331]: The flow_name has a default value in `RaterConfig`.

Comment (Panzy-18) [reply uniflow/flow/config.py:331]: I also notice that, for transform task, every config has a unique name for different model like `TransformOpenAIFlow`, `TransformHuggingFaceFlow`.

**Comment by Panzy-18** [comment]
Find a bug in notebook. 

If we set `batch_size=2` below:
```python
config = RaterForClassificationHuggingfaceConfig(
    model_config=HuggingfaceModelConfig(batch_size=2)
)
client = RaterClient(config)
output = client.run(data)
pprint.pprint(output)
```

It generates wrong output:
```
[{'output': [{'average_score': 0.5,
              'error': 'No errors.',
              'majority_vote': 'yes',
              'response': ['instruction: Evaluate if a given answer is '
                           'appropriate based on the question and the '
                           'context.\n'
                           '            Follow the format of the examples '
                           'below, consisting of context, question, answer, '
                           'explanation and label (you must choose one from '
                           "['Yes', 'No']).\n"
                           'context: The Eiffel Tower, located in Paris, '
                           'France, is one of the most famous landmarks in the '
                           'world. It was constructed in 1889 and stands at a '
                           'height of 324 meters.\n'
                           'question: When was the Eiffel Tower constructed?\n'
                           'answer: The Eiffel Tower was constructed in 1889.\n'
                           'explanation: The answer is consistency to the fact '
                           'that Eiffel Tower was constructed in 1889 '
                           'mentioned in context, so the answer is correct.\n'
                           'label: Yes\n'
                           'context: Photosynthesis is a process used by '
                           'plants to convert light energy into chemical '
                           'energy. This process primarily occurs in the '
                           'chloroplasts of plant cells.\n'
                           'question: Where does photosynthesis primarily '
                           'occur in plant cells?\n'
                           'answer: Photosynthesis primarily occurs in the '
                           'mitochondria of plant cells.\n'
                           'explanation: The context mentions that '
                           'photosynthesis primarily occurs in the '
                           'chloroplasts of plant cells but not mitochondria '
                           'indicated by answer, so the answer is incorrect.\n'
                           'label: No\n'
                           'context: The Pacific Ocean is the largest and '
                           "deepest of Earth's oceanic divisions. It extends "
                           'from the Arctic Ocean in the north to the Southern '
                           'Ocean in the south.\n'
                           'question: What is the largest ocean on Earth?\n'
                           'answer: The largest ocean on Earth is the Pacific '
                           'Ocean.\n'
                           'explanation: The answer is consistent with the '
                           'fact that Pacific Ocean is the largest ocean as '
                           'stated in context, so the answer is correct.\n'
                           'label: Yes',
                           'instruction: Evaluate if a given answer is '
                           'appropriate based on the question and the '
                           'context.\n'
                           '            Follow the format of the examples '
                           'below, consisting of context, question, answer, '
                           'explanation and label (you must choose one from '
                           "['Yes', 'No']).\n"
                           'context: The Eiffel Tower, located in Paris, '
                           'France, is one of the most famous landmarks in the '
                           'world. It was constructed in 1889 and stands at a '
                           'height of 324 meters.\n'
                           'question: When was the Eiffel Tower constructed?\n'
                           'answer: The Eiffel Tower was constructed in 1889.\n'
                           'explanation: The answer is consistency to the fact '
                           'that Eiffel Tower was constructed in 1889 '
                           'mentioned in context, so the answer is correct.\n'
                           'label: Yes\n'
                           'context: Photosynthesis is a process used by '
                           'plants to convert light energy into chemical '
                           'energy. This process primarily occurs in the '
                           'chloroplasts of plant cells.\n'
                           'question: Where does photosynthesis primarily '
                           'occur in plant cells?\n'
                           'answer: Photosynthesis primarily occurs in the '
                           'mitochondria of plant cells.\n'
                           'explanation: The context mentions that '
                           'photosynthesis primarily occurs in the '
                           'chloroplasts of plant cells but not mitochondria '
                           'indicated by answer, so the answer is incorrect.\n'
                           'label: No\n'
                           'context: Shakespeare, a renowned English '
                           'playwright and poet, wrote 39 plays during his '
                           'lifetime. His works include famous plays like '
                           "'Hamlet' and 'Romeo and Juliet'.\n"
                           'question: How many plays did Shakespeare write?\n'
                           'answer: Shakespeare wrote 31 plays.\n'
                           'explanation: The answer is consistent with the '
                           'fact that Shakespeare wrote 39 plays mentioned in '
                           "context, but he didn't write all those plays "
                           'including Hamlet and Romeo and Juliet which are '
                           'famous ones, so the answer is partially correct.\n'
                           'label: Partially'],
              'scores': [1.0, 0.0],
              'votes': ['yes', 'no']}],
  'root': <uniflow.node.Node object at 0x7f43e85ad120>},
 {'output': [{'average_score': 0.0,
              'error': 'No errors.',
              'majority_vote': 'no',
              'response': ['instruction: Evaluate if a given answer is '
                           'appropriate based on the question and the '
                           'context.\n'
                           '            Follow the format of the examples '
                           'below, consisting of context, question, answer, '
                           'explanation and label (you must choose one from '
                           "['Yes', 'No']).\n"
                           'context: The Eiffel Tower, located in Paris, '
                           'France, is one of the most famous landmarks in the '
                           'world. It was constructed in 1889 and stands at a '
                           'height of 324 meters.\n'
                           'question: When was the Eiffel Tower constructed?\n'
                           'answer: The Eiffel Tower was constructed in 1889.\n'
                           'explanation: The answer is consistency to the fact '
                           'that Eiffel Tower was constructed in 1889 '
                           'mentioned in context, so the answer is correct.\n'
                           'label: Yes\n'
                           'context: Photosynthesis is a process used by '
                           'plants to convert light energy into chemical '
                           'energy. This process primarily occurs in the '
                           'chloroplasts of plant cells.\n'
                           'question: Where does photosynthesis primarily '
                           'occur in plant cells?\n'
                           'answer: Photosynthesis primarily occurs in the '
                           'mitochondria of plant cells.\n'
                           'explanation: The context mentions that '
                           'photosynthesis primarily occurs in the '
                           'chloroplasts of plant cells but not mitochondria '
                           'indicated by answer, so the answer is incorrect.\n'
                           'label: No\n'
                           'context: The human brain is an intricate organ '
                           'responsible for intelligence, memory, and '
                           'emotions. It is made up of approximately 86 '
                           'billion neurons.\n'
                           'question: What is the human brain responsible '
                           'for?\n'
                           'answer: The human brain is responsible for '
                           'physical movement.\n'
                           'explanation: The context states that the human '
                           'brain is responsible for intelligence, memory, and '
                           'emotions which are not related with physical '
                           'movement stated by answer, so the answer is '
                           'incorrect.\n'
                           'label: No'],
              'scores': [0.0],
              'votes': ['no']}],
  'root': <uniflow.node.Node object at 0x7f4561fac100>}]

data 0 has label yes and score 0.5
data 1 has label no and score 0.0
```

File: uniflow/flow/config.py [hunk 4]
```diff
@@ -362,6 +361,51 @@ class RaterForClassificationBedrockClaudeConfig(RaterConfig):
...
+        model_config (ModelConfig): Configuration for the huggingeface model. Includes model_name("mistralai/Mistral-7B-Instruct-v0"),
+                                    model_server ("HuggingfaceModelServer"), batch_size (1), neuron (False),
+                                    load_in_4bit (False), load_in_8bit (True)
+        label2score (Dict[str, float]): Mapping of labels to scores, default is {"Yes": 1.0, "No": 0.0}.
+        guided_prompt_template (GuidedPrompt): Template for guided prompts used in rating. Includes instructions
+                                               for rating, along with examples that detail the context, question,
+                                               answer, label, and explanation for each case.
+    """
+
+    model_config: ModelConfig = field(default_factory=HuggingfaceModelConfig())
```
Comment (CambioML) [inline uniflow/flow/config.py:381]: HuggingFace model is refactored in this https://github.com/CambioML/uniflow/pull/110.
- it introduce 
[text](https://github.com/CambioML/uniflow/blob/main/uniflow/flow/config.py#L122-L124) response format and [json](https://github.com/CambioML/uniflow/blob/main/uniflow/flow/config.py#L158-L160) response format to better control huggingface model especially mistral model.
- it also introduce a new argument `response_start_key` for your case, i believe model generated response starting with `explanation` which should be the start key. 

Please up you your config.py properly to address change in https://github.com/CambioML/uniflow/pull/110 and then in your notebook add use case for both text and json mode.

**Comment by CambioML** [comment]
Also, in your comment above https://github.com/CambioML/uniflow/pull/109#issuecomment-1888021544, could you please help summarize what do you mean is wrong?

**Comment by Panzy-18** [comment]
> Also, in your comment above [#109 (comment)](https://github.com/CambioML/uniflow/pull/109#issuecomment-1888021544), could you please help summarize what do you mean is wrong?

Sorry I seem to misunderstand the function. There are three examples as input data, but the output format is like there are only two, and each has a major vote.

File: uniflow/flow/config.py (hunk 1 above)
Comment (CambioML) [reply uniflow/flow/config.py:509]: You bring a good point, I ended up realizing that we do not need these much classes. We should be able to unified `TrnasformOpenAIFlow` with `TransformHuggingFaceFlow` into a unified `TransformFlow` in the future.

File: uniflow/flow/config.py (hunk 4 above)
Comment (CambioML) [reply uniflow/flow/config.py:559]: @Panzy-18 Any progress on addressing this comment?

**Comment by CambioML** [comment]
please pull https://github.com/CambioML/uniflow/pull/116 to fix your issue on [slack](https://cambiomlworkspace.slack.com/archives/C05HZHD6YKB/p1705217068698199)

**Comment by Panzy-18** [comment]
1. what I have done: add huggingface classification config and notebook(both json and text).
2. why the response is still in the wrong format: the model should generate a json object with two fields: `explanation` and `label`, however, its output is:
```
[{'output': [{'average_score': None,
              'error': 'No errors.',
              'majority_vote': None,
              'response': [{'context': 'The answer is consistent with the fact '
                                       'stated in the context that the Pacific '
                                       'Ocean is the largest ocean on Earth.',
                            'question': 'Yes'}],
              'scores': [],
              'votes': [None]}],
  'root': <uniflow.node.Node object at 0x7f77b18ba1a0>},
 {'output': [{'average_score': None,
              'error': 'No errors.',
              'majority_vote': None,
              'response': [{'context': 'The answer is less than the number of '
                                       'plays mentioned in the context, '
                                       'therefore it is an incorrect answer.',
                            'question': 'No.'}],
              'scores': [],
              'votes': [None]}],
  'root': <uniflow.node.Node object at 0x7f78d52775e0>},
 {'output': [{'average_score': None,
              'error': 'No errors.',
              'majority_vote': None,
              'response': [{'context': '{"instruction": "Evaluate if a given '
                                       'answer is appropriate based on the '
                                       'question and the '
                                       'context.\\n            Follow the '
                                       'format of the examples below, '
                                       'consisting of context, question, '
                                       'answer, explanation and label (you '
                                       "must choose one from ['Yes', "
                                       "'No']).\\n\\nEnsure the response is in "
                                       'json.", "few_shot_prompt": '
                                       '[{"context": "The Eiffel Tower, '
                                       'located in Paris, France, is one of '
                                       'the most famous landmarks in the '
                                       'world. It was constructed in 1889 and '
                                       'stands at a height of 324 meters.", '
                                       '"question": "When was the Eiffel Tower '
                                       'constructed?", "answer": "The Eiffel '
                                       'Tower was constructed in 1889.", '
                                       '"explanation": "The answer is '
                                       'consistency to the fact that Eiffel '
                                       'Tower was constructed in 1889 '
                                       'mentioned in context, so the answer is '
                                       'correct.", "label": "Yes"}, '
                                       '{"context": "Photosynthesis is a '
                                       'process used by plants to convert '
                                       'light energy into chemical energy. '
                                       'This process primarily occurs in the '
                                       'chloroplasts of plant cells.", '
                                       '"question": "Where does photosynthesis '
                                       'primarily occur in plant cells?", '
                                       '"answer": "Photosynthesis primarily '
                                       'occurs in the mitochondria of plant '
                                       'cells.", "explanation": "The context '
                                       'mentions that photosynthesis primarily '
                                       'occurs in the chloroplasts of plant '
                                       'cells but not mitochondria indicated '
                                       'by answer, so the answer is '
                                       'incorrect.", "label": "No"}, '
                                       '{"context": "The human brain is an '
                                       'intricate organ responsible for '
                                       'intelligence, memory, and emotions. It '
                                       'is made up of approximately 86 billion '
                                       'neurons.", "question": "What is the '
                                       'human brain responsible for?", '
                                       '"answer": "The human brain is '
                                       'responsible for physical movement."}]}',
                            'question': 'The answer provided in the third '
                                        'example is not consistent with the '
                                        'facts stated in the context. The '
                                        'human brain is indeed responsible for '
                                        'intelligence, memory, and emotions as '
                                        'stated in the context, not physical '
                                        'movement. Therefore, the label should '
                                        "be 'No'. Here's the corrected JSON:\n"
                                        '{\n'
                                        '"context": "The human brain is an '
                                        'intricate organ responsible for '
                                        'intelligence, memory, and emotions. '
                                        'It is made up of approximately 86 '
                                        'billion neurons.",\n'
                                        '"question": "What is the human brain '
                                        'responsible for?",\n'
                                        '"answer": "The human brain is '
                                        'responsible for physical movement.",\n'
                                        '"explanation": "The answer is '
                                        'inconsistent with the information '
                                        'provided in the context which states '
                                        'that the human brain is responsible '
                                        'for intelligence, memory, and '
                                        'emotions.",\n'
                                        '"label": "No"\n'
                                        '}'}],
              'scores': [],
              'votes': [None]}],
  'root': <uniflow.node.Node object at 0x7f78d5277fd0>}]
```
We can see model seems to generate a json object with `context` and `question`. In the former 2 examples, model in fact generate the expected CoT explanation and label. In the last example, model generate a complex nested json object.

**Comment by Panzy-18** [comment]
I notice in `HuggingfaceModelServer`, `_preprocess()` add some special token before and after instruction. For input example
```
("The Pacific Ocean is the largest and deepest of Earth's oceanic divisions. It extends from the Arctic Ocean in the north to the Southern Ocean in the south.",
     "What is the largest ocean on Earth?",
     "The largest ocean on Earth is the Pacific Ocean."), # correct
```
It pass this string to LLM:
```
['<s>[INST] {"instruction": "Evaluate if a given answer is appropriate based on the question and the context.\\n            Follow the format of the examples below, consisting of context, question, answer, explanation and label (you must choose one from [\'Yes\', \'No\']).\\n\\nEnsure the response is in json.", "few_shot_prompt": [{"context": "The Eiffel Tower, located in Paris, France, is one of the most famous landmarks in the world. It was constructed in 1889 and stands at a height of 324 meters.", "question": "When was the Eiffel Tower constructed?", "answer": "The Eiffel Tower was constructed in 1889.", "explanation": "The answer is consistency to the fact that Eiffel Tower was constructed in 1889 mentioned in context, so the answer is correct.", "label": "Yes"}, {"context": "Photosynthesis is a process used by plants to convert light energy into chemical energy. This process primarily occurs in the chloroplasts of plant cells.", "question": "Where does photosynthesis primarily occur in plant cells?", "answer": "Photosynthesis primarily occurs in the mitochondria of plant cells.", "explanation": "The context mentions that photosynthesis primarily occurs in the chloroplasts of plant cells but not mitochondria indicated by answer, so the answer is incorrect.", "label": "No"}, {"context": "The Pacific Ocean is the largest and deepest of Earth\'s oceanic divisions. It extends from the Arctic Ocean in the north to the Southern Ocean in the south.", "question": "What is the largest ocean on Earth?", "answer": "The largest ocean on Earth is the Pacific Ocean."}]} [/INST]\nexplanation: ']
```

This may lead model to generate a json object with the mutual fields shown in  what was wrapped by `[INST]` and `[/INST]`,  therefore to generate a json object {"context":..., "question": ...}.

**Comment by Panzy-18** [comment]
I modify the string passed to LLM as below:
```
'<s>[INST] {"instruction": "Evaluate if a given answer is appropriate based on the question and the context.\\n            Follow the format of the examples below, consisting of context, question, answer, explanation and label (you must choose one from [\'Yes\', \'No\']).\\n\\nEnsure the response is in json.", "few_shot_prompt": [{"context": "The Eiffel Tower, located in Paris, France, is one of the most famous landmarks in the world. It was constructed in 1889 and stands at a height of 324 meters.", "question": "When was the Eiffel Tower constructed?", "answer": "The Eiffel Tower was constructed in 1889.", "explanation": "The answer is consistency to the fact that Eiffel Tower was constructed in 1889 mentioned in context, so the answer is correct.", "label": "Yes"}, {"context": "Photosynthesis is a process used by plants to convert light energy into chemical energy. This process primarily occurs in the chloroplasts of plant cells.", "question": "Where does photosynthesis primarily occur in plant cells?", "answer": "Photosynthesis primarily occurs in the mitochondria of plant cells.", "explanation": "The context mentions that photosynthesis primarily occurs in the chloroplasts of plant cells but not mitochondria indicated by answer, so the answer is incorrect.", "label": "No"}]} [/INST]\n {"context": "The Pacific Ocean is the largest and deepest of Earth\'s oceanic divisions. It extends from the Arctic Ocean in the north to the Southern Ocean in the south.", "question": "What is the largest ocean on Earth?", "answer": "The largest ocean on Earth is the Pacific Ocean.", "explanation":'
```

And get the output in `_postprocess()`:

```
[{'generated_text': '<s>[INST] {"instruction": "Evaluate if a given answer is appropriate based on the question and the context.\\n            Follow the format of the examples below, consisting of context, question, answer, explanation and label (you must choose one from [\'Yes\', \'No\']).\\n\\nEnsure the response is in json.", "few_shot_prompt": [{"context": "The Eiffel Tower, located in Paris, France, is one of the most famous landmarks in the world. It was constructed in 1889 and stands at a height of 324 meters.", "question": "When was the Eiffel Tower constructed?", "answer": "The Eiffel Tower was constructed in 1889.", "explanation": "The answer is consistency to the fact that Eiffel Tower was constructed in 1889 mentioned in context, so the answer is correct.", "label": "Yes"}, {"context": "Photosynthesis is a process used by plants to convert light energy into chemical energy. This process primarily occurs in the chloroplasts of plant cells.", "question": "Where does photosynthesis primarily occur in plant cells?", "answer": "Photosynthesis primarily occurs in the mitochondria of plant cells.", "explanation": "The context mentions that photosynthesis primarily occurs in the chloroplasts of plant cells but not mitochondria indicated by answer, so the answer is incorrect.", "label": "No"}]} [/INST]\n {"context": "The Pacific Ocean is the largest and deepest of Earth\'s oceanic divisions. It extends from the Arctic Ocean in the north to the Southern Ocean in the south.", "question": "What is the largest ocean on Earth?", "answer": "The largest ocean on Earth is the Pacific Ocean.", "explanation":  "The answer is consistent with the fact stated in the context, therefore the answer is correct.", "label": "Yes"}'}]
```

This output is more easily to be parsed. We can ignore the instruction(prompt, few_shot..) wrapped by `[INST]` and `[/INST]`, the string will be:

```
[{'generated_text': '<s>\n {"context": "The Pacific Ocean is the largest and deepest of Earth\'s oceanic divisions. It extends from the Arctic Ocean in the north to the Southern Ocean in the south.", "question": "What is the largest ocean on Earth?", "answer": "The largest ocean on Earth is the Pacific Ocean.", "explanation":  "The answer is consistent with the fact stated in the context, therefore the answer is correct.", "label": "Yes"}'}]
```

File: uniflow/flow/config.py [hunk 5]
```diff
@@ -540,6 +539,57 @@ class RaterForClassificationBedrockClaudeConfig(RaterConfig):
...
+        guided_prompt_template (GuidedPrompt): Template for guided prompts used in rating. Includes instructions
+                                               for rating, along with examples that detail the context, question,
+                                               answer, label, and explanation for each case.
+    """
+
+    model_config: ModelConfig = field(
+        default_factory=HuggingfaceModelConfig(
+            response_start_key="explanation",
+            response_format={"type": "text"},
+            batch_size=1
```
Comment (CambioML) [inline uniflow/flow/config.py:563]: I think you miss a comma here which trigger a black failure during build time. If you have not do so, follow https://www.notion.so/goldpiggy/Python-Linter-and-formatter-Setup-30fb3b81f0904af889832e4c697c5ec9?pvs=4 to properly setup your dev space.

**Comment by CambioML** [comment]
@Panzy-18 This is a good catch https://github.com/CambioML/uniflow/pull/109#issuecomment-1888021544!

**Comment by CambioML** [comment]
https://github.com/CambioML/uniflow/pull/109#issuecomment-1890888262, https://github.com/CambioML/uniflow/pull/109#issuecomment-1890890131, and https://github.com/CambioML/uniflow/pull/109#issuecomment-1890892375 are great analysis. Thanks for the details description and propose your solution. per your request, I fixed it in this https://github.com/CambioML/uniflow/pull/119. Could you please pull the latest code and try again. Let me know whether it fixed your issues. Meanwhile, let's use `batch_size = 1` for now. We will fix it in a separated PR.

File: uniflow/flow/config.py (hunk 5 above)
Comment (Panzy-18) [reply uniflow/flow/config.py:563]: sorry I forget to press command+s before push

**Review by CambioML — APPROVED** [review]
LGTM.

**Pull request merged**